rflx/model/message.py

```diff
@@ -1569,6 +1569,52 @@ def proven(self, skip_proof: bool = False) -> "DerivedMessage":
         return copy(self)
 
 
+def replace_message_attributes(
+    message: AbstractMessage,
+) -> Tuple[AbstractMessage, List[Optional[Location]]]:
+    """
+    Replace all occurences of the message attribute Message'Size by the equivalent term
+    Message'Last - Message'First + 1. Then replace all occurences of the message attribute
+    Message'First by a reference to the first field of the message. Return the new message and
+    a list of locations where Message'Last or Message'Size have been used.
+    """
+    first_field = message.outgoing(INITIAL)[0].target
+    locations = []
+
+    def replace(expression: expr.Expr) -> expr.Expr:
+        if (
+            not isinstance(expression, expr.Attribute)
+            or not isinstance(expression.prefix, expr.Variable)
+            or expression.prefix.name != "Message"
+        ):
+            return expression
+        if isinstance(expression, expr.First):
+            return expr.First(expr.ID(first_field.identifier, location=expression.location))
+        locations.append(expression.location)
+        if isinstance(expression, expr.Last):
+            return expression
+        if isinstance(expression, expr.Size):
+            return expr.Add(
+                expr.Sub(expr.Last("Message"), expr.First(first_field.identifier)),
+                expr.Number(1),
+                location=expression.location,
+            )
+        raise NotImplementedError
```

Comment:
`NotImplementedError` is not the right exception here (cf. https://docs.python.org/3/library/exceptions.html#NotImplementedError).

Reply:
Changed to `TypeError` in 877fd78d.

Reply:
An alternative would be an `assert False` here (at least this is what is currently used in most similar cases). Otherwise, an additional unit test is needed.

Reply:
That's true. I don't even know how to come up with such a test. Changed in 643c841f.